exporters/etw/README.md

```diff
@@ -0,0 +1,201 @@
+use them to debug an application or to determine where performance issues are occurring in the application.
+
+OpenTelemetry C++ SDK ETW exporter allows the code instrumented using OpenTelemetry API to forward events
+to out-of-process ETW listener, for subsequent data recording or forwarding to alternate pipelines and
+flows. Windows Event Tracing infrastructure is available to any vendor or application being deployed on
+Windows.
+
+## API support
+
+These are the features planned to be supported by ETW exporter:
+
+- [x] OpenTelemetry Tracing API and SDK headers are **stable** and moving towards GA.
+- [ ] OpenTelemetry Logging API is work-in-progress, pending implementation of [Latest Logging API spec here](https://github.com/open-telemetry/oteps/pull/150)
+- [ ] OpenTelemetry Metrics API is not implemented yet.
+
+Implementation of OpenTelemetry C++ SDK ETW exporter on Windows OS is `header only` :
+
+- full definitions of all macros, functions and classes comprising the library are visible to the compiler
+in a header file form.
+- implementation does not need to be separately compiled, packaged and installed in order to be used.
+
+All that is required is to point the compiler at the location of the headers, and then `#include` the header
+files into the application source. Compiler's optimizer can do a much better job when all the library's
+source code is available. Several options below may be turned on to optimize the code with the usage of
+standard C++ library, Microsoft Guidelines Support library, Google Abseil Variant library. Or enabling
+support for non-standard features, such as 8-bit byte arrays support that enables performance-efficient
+representation of binary blobs on ETW wire.
+
+## Example project
+
+The following include directories are required, relative to the top-level source tree of OpenTelemetry C++ SDK:
+
+- api/include/
+- exporters/etw/include/
+- sdk/include/
+
+Code that instantiates ETW TracerProvider, subsequently obtaining a Tracer bound to `OpenTelemetry-ETW-Provider`,
+and emitting a span named `MySpan` with attributes on it, as well as `MyEvent` within that span.
+
+```cpp
+
+#include <map>
+#include <string>
+
+#include "opentelemetry/exporters/etw/etw_tracer_exporter.h"
+
+using namespace OPENTELEMETRY_NAMESPACE;
+using namespace opentelemetry::exporter::ETW;
+
+// Supply unique instrumentation name (ETW Provider Name) here:
+std::string providerName = "OpenTelemetry-ETW-Provider";
+
+exporter::ETW::TracerProvider tp;
+
+int main(int argc, const char* argv[])
+{
+  // Obtain a Tracer object for instrumentation name.
+  // Each Tracer is associated with unique TraceId.
+  auto tracer = tp.GetTracer(providerName, "TLD");
+
+  // Properties is a helper class in ETW namespace that is otherwise compatible
+  // with Key-Value Iterable accepted by OpenTelemetry API. Using Properties
+  // should enable more efficient data transfer without unnecessary memcpy.
```

Comment:
That's a bit misleading in the context of this example. As far as I can see, none of the usages of `Properties` here is more efficient than what the API provides out of the box. There's no `memcpy` involved in this case:
```c++
span->AddEvent(eventName, {
    {"uint32Key", (uint32_t)1234},
    {"uint64Key", (uint64_t)1234567890},
    {"strKey", "someValue"}
});
```

Reply:
The answer - really depends. When we need to decorate the properties with additional attributes, such as - passing non-const properties, append some common properties... Maybe unique to custom `TracerProvider` ), for example - some sort of custom correlation technique that needs to be appended on every record. Let's say, some custom field that needs to be called `cV` and stamped on every `Properprties`. Then we could have reused the container that was created by the customer code once, then passing it in essence by reference on one thread (not requiring a transform to `Recordable` ), and not needing the key-value iterable for each key-value. That'd be more efficient than for each key-value copy, then transform to recordable, etc.

Reply:
I'll leave this code reference as an example where accepting the collection, then passing this **original collection** around by reference, decorating it with additional contents, stripping unneded fields, etc. would be more efficient than trying to construct a new object (or a copy of an object) from initializer list. I think I understand your concern that I should've elaborated on exactly what scenarios would (in future) may benefit from passing in the original collection. This should work great for a header-only library, as we later on may completely bypass the copy/iteration/creation of a new object from initializer list, operating on original customer-code created object.

We mostly follow this pattern in this library elsewhere:
https://github.com/microsoft/cpp_client_telemetry/blob/master/lib/api/Logger.cpp
And it's been working well.

Reply:
Here is another scenario: let's say your `collection does not change` and contains something like `appName=x,status=Failed,scenario=FileOpen`. You'd prepare it once in a static initializer. Then you pass it as a reference to `Properties`, and never reconstruct it instead of constructing it from Key-Value Iterable every time. It appears like this would be more optimal (at least for a statically linked code where there is no passing across ABI boundary needed). But maybe I'm wrong - we need to take a look at both and run some benchmarks. I can remove the performance comment if it sounds wrong to you.

Reply:
Can we add the specific example/scenario as an example in this doc or in unit-test ( and refer it here ) to justify the performance statement. As the current api/sdk implementation, the only place where `memcpy` MAY happen is during Recordable::AddEvent() execution, but that would depend how exporter implements this logic.

Reply:
> the only place where memcpy MAY happen is during Recordable::AddEvent() execution

Well, that is the most commonly invoked function, actually.

Reply:
I'm pushing an update to my code PR to explain this further.